tsl/src/compression/api.c

```diff
@@ -892,3 +896,584 @@ tsl_recompress_chunk_wrapper(Chunk *uncompressed_chunk)
+									   0 /*flags*/,
+									   attno,
+									   BTEqualStrategyNumber,
+									   InvalidOid,
+									   current_segment[seg_col]->segment_info->collation,
+									   &current_segment[seg_col]->segment_info->eq_fn,
+									   val);
+	}
+
+	heapScan =
+		table_beginscan(uncompressed_chunk_rel, GetLatestSnapshot(), nsegmentby_cols, scankey);
+
+	for (uncompressed_tuple = heap_getnext(heapScan, ForwardScanDirection);
+		 uncompressed_tuple != NULL;
+		 uncompressed_tuple = heap_getnext(heapScan, ForwardScanDirection))
+	{
+		TupleTableSlot *heap_tuple_slot =
+			MakeTupleTableSlot(uncompressed_rel_tupdesc, &TTSOpsHeapTuple);
+
+		ExecStoreHeapTuple(uncompressed_tuple, heap_tuple_slot, false);
+		slot_getallattrs(heap_tuple_slot);
+		tuplesort_puttupleslot(segment_tuplesortstate, heap_tuple_slot);
+		/* simple_heap_delete since we don't expect concurrent updates, have exclusive lock on the
+		 * relation */
+		simple_heap_delete(uncompressed_chunk_rel, &uncompressed_tuple->t_self);
+		ExecDropSingleTupleTableSlot(heap_tuple_slot);
+	}
+	table_endscan(heapScan);
+	pfree(scankey);
+}
+/*
+ * This function will replace the procedure for recompress_chunk
+ * It will keep the compressed chunk around, only decompressing the
+ * batches that are affected by the addition of newer data.
+ * If more batches need to be created, it will create those and update
+ * the meta_sequence_number of the following ones.
+ *
+ * 0 uncompressed_chunk_id REGCLASS
+ * 1 if_not_compressed BOOL = false
+ */
+Datum
+tsl_recompress_chunk_segmentwise(PG_FUNCTION_ARGS)
+{
+	Oid uncompressed_chunk_id = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	bool if_not_compressed =
+		PG_ARGISNULL(1) ? false : PG_GETARG_BOOL(1); // probably don't need this
+
+	TS_PREVENT_FUNC_IF_READ_ONLY();
+	Chunk *uncompressed_chunk = ts_chunk_get_by_relid(uncompressed_chunk_id, true);
+
+	if (NULL == uncompressed_chunk)
+		elog(ERROR, "unknown chunk id %d", uncompressed_chunk_id);
+
+	/* need to push down to data nodes if this is an access node */
+	if (uncompressed_chunk->relkind == RELKIND_FOREIGN_TABLE)
+	{
+		if (!recompress_remote_chunk(fcinfo, uncompressed_chunk))
+			PG_RETURN_NULL();
+
+		/* TODO: probably redundant? */
+		ts_chunk_set_compressed_chunk(uncompressed_chunk, INVALID_CHUNK_ID);
+
+		PG_RETURN_OID(uncompressed_chunk_id);
+	}
+
+	int32 status = uncompressed_chunk->fd.status;
+
+	if (status == CHUNK_STATUS_DEFAULT)
+		elog(ERROR, "call compress_chunk instead of recompress_chunk");
+	if (status == CHUNK_STATUS_COMPRESSED)
+	{
+		int elevel = if_not_compressed ? NOTICE : ERROR;
+		elog(elevel,
+			 "nothing to recompress in chunk %s.%s",
+			 uncompressed_chunk->fd.schema_name.data,
+			 uncompressed_chunk->fd.table_name.data);
+	}
+
+	/*
+	 * only proceed if status in (3, 9, 11)
+	 * 1: compressed
+	 * 2: compressed_unordered
+	 * 4: frozen
+	 * 8: compressed_partial
+	 */
+	if (status != 3 && status != 9 && status != 11)
```

Comment:
I think the readability of the condition could be improved if you use `ts_chunk_is_unordered` / `ts_chunk_is_compressed` / `ts_chunk_is_partial` here.

Reply:
done